instrumentation/github.com/aws/aws-sdk-go/service/otels3/s3.go

```diff
@@ -0,0 +1,201 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otels3
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+
+	"go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go/service/config"
+	"go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-sdk-go/service/helper"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/metric"
+	"go.opentelemetry.io/otel/api/trace"
+)
+
+var instrumentationName = "github.com/aws/aws-sdk-go/aws/service/s3"
+
+type instrumentedS3 struct {
+	s3iface.S3API
+	tracer                   trace.Tracer
+	meter                    metric.Meter
+	propagators              otel.TextMapPropagator
+	counters                 *counters
+	recorders                *recorders
+	spanCorrelationInMetrics bool
+}
+
+type counters struct {
+	operation metric.Int64Counter
+}
+
+type recorders struct {
+	operationDuration metric.Float64ValueRecorder
+}
+
+// PutObjectWithContext invokes the PutObjectWithContext function with tracing instrumented
+func (s *instrumentedS3) PutObjectWithContext(ctx aws.Context, input *s3.PutObjectInput, opts ...request.Option) (*s3.PutObjectOutput, error) {
+	startTime := time.Now()
+	destination := aws.StringValue(input.Bucket)
+	attrs := createAttributes(destination, operationPutObject)
+
+	spanCtx, span := s.tracer.Start(
+		ctx,
+		fmt.Sprintf("%s.%s", destination, operationPutObject),
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(attrs...),
+	)
+	defer span.End()
```

Comment:
By deferring this to the function return it will include timing for metric recording. Instead, saving the time the downstream call returns to a variable and passing to `End` will keep this accurate.